src/Processors/QueryPlan/Optimizations/removeRedundantDistinct.cpp

```diff
@@ -0,0 +1,147 @@
+#include <memory>
+#include <Interpreters/ActionsDAG.h>
+#include <Processors/QueryPlan/DistinctStep.h>
+#include <Processors/QueryPlan/ExpressionStep.h>
+#include <Processors/QueryPlan/FilterStep.h>
+#include <Processors/QueryPlan/IntersectOrExceptStep.h>
+#include <Processors/QueryPlan/JoinStep.h>
+#include <Processors/QueryPlan/Optimizations/Optimizations.h>
+#include <Processors/QueryPlan/UnionStep.h>
+#include <Common/logger_useful.h>
+#include <Common/typeid_cast.h>
+
+namespace DB::QueryPlanOptimizations
+{
+
+namespace
+{
+    constexpr bool debug_logging_enabled = false;
+    void logActionsDAG(const String & prefix, const ActionsDAGPtr & actions)
+    {
+        if constexpr (debug_logging_enabled)
+        {
+            LOG_DEBUG(&Poco::Logger::get("redundantDistinct"), "{}: {}", prefix, actions->dumpDAG());
+        }
+    }
+
+    std::set<std::string_view> getDistinctColumns(const DistinctStep * distinct)
+    {
+        /// find non-const columns in DISTINCT
+        const ColumnsWithTypeAndName & distinct_columns = distinct->getOutputStream().header.getColumnsWithTypeAndName();
+        std::set<std::string_view> non_const_columns;
+        for (const auto & column : distinct_columns)
+        {
+            if (!isColumnConst(*column.column))
+                non_const_columns.emplace(column.name);
+        }
+        return non_const_columns;
+    }
+
+    bool canRemoveDistinct(const QueryPlan::Node * distinct_node)
+    {
+        const DistinctStep * distinct_step = typeid_cast<DistinctStep *>(distinct_node->step.get());
+        chassert(distinct_step);
+
+        std::vector<ActionsDAGPtr> dag_stack;
+        const DistinctStep * inner_distinct_step = nullptr;
+        const QueryPlan::Node * node = distinct_node;
+        while (!node->children.empty())
+        {
+            const IQueryPlanStep * current_step = node->step.get();
+
+            /// don't try to remove DISTINCT after union/join/intersect/except
+            if (typeid_cast<const UnionStep *>(current_step) || typeid_cast<const JoinStep *>(current_step)
```

Comment:
I think we must skip all the steps with > 1 inputs.
Also, I would prefer to have allow list logic. Have a list of steps which we support. E.g. here I think we should skip `ArrayJoinStep`

Reply:
Let's go over setps:

- Aggregating, Rollup, Cube, TotalsHaving: probably ok. But it will work only if distinct columns == GROUP BY keys. I would better handle this case separately: remove distinct if it is after group by and distinct columns include GROUP BY key.
- ArrayJoin, FillingStep : cannot remove distinct cause it can copy cows
- Expressin, Filter : need additionally check `!hasArrayJoin`
- Limit, Limit by: ok
- Sorting: probably ok
- Window: technically should be ok cause it does not generate new rows, but maybe double-check ...

Reply:
From naming the optimization is sound like generic one, but actually, within this PR, the focus was on removing DISTINCT due to duplication of DISTINCT clauses. It means that if there are two DISTINCT steps on the same columns, we need to ensure that there are _no_ steps between them which can generate new rows.
So, I've added corresponding checks:
- ArrayJoin, FillingStep : cannot remove distinct because it can copy cows
- Expressin, Filter : need additionally check !hasArrayJoin

More generic opportunities to remove DISTINCT like with aggregation can be done separately.